GQI Caching page (#5042)
SkylineCommunications/dataminer-docs: develop/devguide/GQI Extensions/Best Practices/GQI_Extensions_Caching.md

- **Manual reset (cache breaker)**: Use a file, flag, or external signal to clear or refresh the cache on demand. This is useful for administrative control when data changes are event-driven.
- **Hybrid approach**: Combine timeouts with manual reset to balance predictable refresh with flexibility.
- **Invalidate per security group**: If only certain groups are affected, clear cache entries selectively rather than globally.

Carefully plan invalidation frequency. Excessively frequent invalidation negates the benefits of caching, while infrequent invalidation risks serving stale or incorrect data.

### Concurrency

In multi-user or multi-threaded environments, cache methods may be called concurrently by different threads. Without proper synchronization, this can cause race conditions, redundant data fetching, or an inconsistent cache state.

To handle concurrency safely:

- Use thread-safe collections such as `ConcurrentDictionary` for shared caches.
- Protect critical sections (e.g. cache initialization or updates) with locks (`lock` statement or other synchronization primitives).

> - Double-check locking patterns help ensure initialization/update occurs only once per cache cycle, minimizing duplicate work.

Reviewer: This sentence isn't entirely clear. Is this what you mean?
```suggestion
- Double-check whether locking patterns help ensure that initialization/updating occurs only once per cache cycle, minimizing duplicate work.
```

Reviewer 2: ["Double-checked locking"](https://en.wikipedia.org/wiki/Double-checked_locking) is a pattern.

So it should be something like:
```suggestion
- Use [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) to ensure that initializing/updating occurs only once per cache cycle, minimizing duplicate work.
```

Reviewer 2: Not really, it's a concept that applies to programming in general. There is a paper and a book that explains it in detail (see e.g. https://www.dre.vanderbilt.edu/~schmidt/PDF/DC-Locking.pdf) and I found a section in https://learn.microsoft.com/en-us/archive/msdn-magazine/2005/october/understanding-low-lock-techniques-in-multithreaded-apps but I think that misses the point of explaining it to someone who's not familiar with it.

- Accuracy must always be real-time (e.g. live status, alarms).
- The dataset is very small and fast to compute (caching adds unnecessary complexity).

## How To Add Caching

Adding caching to a GQI extension is not natively supported by the framework and must be implemented manually. The most common approach is to use in-memory caching with static variables (for example a `ConcurrentDictionary`) to share data between queries and operators within the same precompiled script.

> Because each precompiled script runs in its own process, this static cache is process-local. It won’t be shared across multiple compiled scripts or across DataMiner agents. If you need to share cached data beyond a single process, you must use a different approach, for example file-based caching.

Author: I was thinking the same before commiting I had precompiled scripts/libraries. but then changed it to scripts as I thought that was the most common way of referencing it. But I like the Extension Libraries better.

## Key Considerations When Caching

> - [**Cache lifetime**](#cache-lifetime): Know which objects can safely be cached and which framework dependencies must never be stored statically.

Author: No problem in changing them. How would you order them?

> - [**Cache lifetime**](#cache-lifetime): Know which objects can safely be cached and which framework dependencies must never be stored statically.

Reviewer 3: I have no specific order in mind, now it's not part of the other entries of your list. So I believe you just need to put the  "When adding caching to GQI extensions, there are several aspects you should evaluate:" line above this.
